### PR TITLE
Allow to asynchronously notify extensions of no-restart changes

### PR DESCRIPTION
**Pull request by ia3andy**
This allows a clean way to notify extensions of no-restart changes like there: https://github.com/quarkiverse/quarkus-web-bundler/pull/217:
```java
RuntimeUpdatesProcessor.INSTANCE
                .notifyExtensions(noRestartChanges)
```

This also allows to reset the compileProblem asynchronously.

For context, when the change is already handled/compiled by the underlying process, triggering a doScan() is not necessary and could even be an issue. Still, extensions needs to be notified of the change.

### CODE REVIEW

**Comment by ia3andy** [comment]
Tested with web-bundler

File: core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java [hunk 1]
```diff
@@ -585,6 +587,27 @@ public boolean doScan(boolean userInitiated, boolean forceRestart) {
...
+     *
+     * @param noRestartChangedFiles the Set of changed files
+     */
+    public void notifyExtensions(Set<String> noRestartChangedFiles) {
+        if (lastStartIndex == null) {
+            // we don't notify extensions if the application never started
+            return;
+        }
+        try {
+            scanLock.lock();
```
Comment (FroMage) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java:601]: Shouldn't it be `lock` before the `try` like all other such cases?

**Review by FroMage — APPROVED** [review]
Looks good.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java (hunk 1 above)
Comment (ia3andy) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java]: Done, I also added `codeGenLock`

**Review by FroMage — APPROVED** [review]

File: core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java [hunk 2]
```diff
@@ -585,6 +587,30 @@ public boolean doScan(boolean userInitiated, boolean forceRestart) {
...
+     *
+     * @param noRestartChangedFiles the Set of changed files
+     */
+    public void notifyExtensions(Set<String> noRestartChangedFiles) {
+        if (lastStartIndex == null) {
+            // we don't notify extensions if the application never started
+            return;
+        }
+        scanLock.lock();
+        codeGenLock.lock();
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java:601]: Why the new locks?

Comment (gsmet) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java:601]: For the context, I want to make sure this won't cause problems as you asked for a backport.

**Pull request merged**